src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts

```diff
@@ -450,11 +452,13 @@ export class TerminalTaskSystem implements ITaskSystem {
 					startStopProblemMatcher.done();
 					startStopProblemMatcher.dispose();
 					registeredLinkMatchers.forEach(handle => terminal.deregisterLinkMatcher(handle));
-					if (processStartedSignaled) {
-						this._onDidStateChange.fire(TaskEvent.create(TaskEventKind.ProcessEnded, task, exitCode));
-					}
-					this._onDidStateChange.fire(TaskEvent.create(TaskEventKind.Inactive, task));
-					this._onDidStateChange.fire(TaskEvent.create(TaskEventKind.End, task));
+					processStartedSignaled.then((success: boolean) => {
+						if (success) {
+							this._onDidStateChange.fire(TaskEvent.create(TaskEventKind.ProcessEnded, task, exitCode));
+						}
+						this._onDidStateChange.fire(TaskEvent.create(TaskEventKind.Inactive, task));
```

Comment:
If TaskEventKind.Inactive is fired before TaskEventKind.ProcessEnded, then $onDidEndTaskProcess will not fire because before it is fired we check for active tasks. If TaskEventKind.Inactive is moved out of the then block then we still have this bug. I can instead move the TaskEventKind.Start and TaskEventKind.Active into the terminal.ProcessReady.then block above.